lib/logstash/inputs/redis.rb

```diff
@@ -73,38 +94,79 @@ def register
       @data_type = 'list'
     end
 
-    if not @key or not @data_type
+    if !@key || !@data_type
       raise RuntimeError.new(
         "Must define queue, or key and data_type parameters"
       )
     end
     # end TODO
 
-    @logger.info("Registering Redis", :identity => identity)
+    @redis_builder ||= method(:internal_redis_builder)
+
+    # just switch on data_type once
+    if @data_type == 'list' || @data_type == 'dummy'
+      @run_method = method(:list_runner)
+      @teardown_method = method(:list_teardown)
+    elsif @data_type == 'channel'
+      @run_method = method(:channel_runner)
+      @teardown_method = method(:subscribe_teardown)
+    elsif @data_type == 'pattern_channel'
+      @run_method = method(:pattern_channel_runner)
+      @teardown_method = method(:subscribe_teardown)
+    end
+
+    # TODO(sissel, boertje): set @identity directly when @name config option is removed.
+    @identity = @name != 'default' ? @name : "#{@redis_url} #{@data_type}:#{@key}"
+    @logger.info("Registering Redis", :identity => @identity)
   end # def register
 
-  # A string used to identify a Redis instance in log messages
-  # TODO(sissel): Use instance variables for this once the @name config
-  # option is removed.
-  private
-  def identity
-    @name || "#{@redis_url} #{@data_type}:#{@key}"
+  def run(output_queue)
+    @run_method.call(output_queue)
+  rescue LogStash::ShutdownSignal
+    # ignore and quit
+  end # def run
+
+  def teardown
+    @shutdown_requested = true
+    @teardown_method.call
   end
 
+  # private methods -----------------------------
   private
-  def connect
-    redis = Redis.new(
+
+  def batched?
+    @batch_count > 1
+  end
+
+  # private
+  def is_list_type?
+    @data_type == 'list'
+  end
+
+  # private
+  def redis_params
+    {
       :host => @host,
       :port => @port,
       :timeout => @timeout,
       :db => @db,
       :password => @password.nil? ? nil : @password.value
-    )
-    load_batch_script(redis) if @data_type == 'list' && (@batch_count > 1)
-    return redis
+    }
+  end
+
+  # private
+  def internal_redis_builder
```

Comment:
I had to add two ways of getting an external Redis instance into the input.

1. Direct injection with the "use_redis" method. Stores an instance, mock or real, of redis
2. Indirect injection with the "add_external_redis_builder" method.  Stores a 'callable' (anything that responds to "call" - lambda, proc, block or ruby class with a call method) that returns a mock or real redis instance.

You will see in the "subscribe_teardown" method there is a "publish_poison_message".
This method needs a new redis instance to publish a poison message because if one uses the current one the publish method will block - waiting on mutex that will never be unlocked.

But the channel specs need a public "new_redis_instance" method to publish real messages. One could use ``` Redis.new ``` but when (in the input) the redis connection details are not the default then one needs to say "input, give me an instance of the client that you are configured to use" and that instance might be a mock or real depending on other setups for the test.

So internal v external.  Maybe "injected" instead of "external" and "defined" instead of "internal"
e.g. "add_injected_redis_builder" v "add_external_redis_builder" and "defined_redis_builder" v "internal_redis_builder".

Re: "builder", it really is a Factory but in the Ruby world, the word "factory" make most people feel sick :-)